src/modules/apps/manage/EditAppModal.tsx

```diff
@@ -0,0 +1,191 @@
+/**
+ * Copyright 2024 IBM Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Project } from '@/app/api/projects/types';
+import { Modal } from '@/components/Modal/Modal';
+import { SettingsFormGroup } from '@/components/SettingsFormGroup/SettingsFormGroup';
+import { ModalProps, useModal } from '@/layout/providers/ModalProvider';
+import {
+  Button,
+  InlineLoading,
+  InlineNotification,
+  ModalBody,
+  ModalFooter,
+  ModalHeader,
+  TextArea,
+  TextInput,
+} from '@carbon/react';
+import { useCallback, useEffect, useId } from 'react';
+import { SubmitHandler, useForm } from 'react-hook-form';
+import classes from './EditAppModal.module.scss';
+import { useModalControl } from '@/layout/providers/ModalControlProvider';
+import { Artifact } from '../types';
+import { useSaveApp } from '../hooks/useSaveApp';
+import { ArtifactCreateBody } from '@/app/api/artifacts/types';
+import isEmpty from 'lodash/isEmpty';
+
+interface FormValues {
+  icon: string;
+  name: string;
+  description?: string;
+}
+
+interface Props extends ModalProps {
+  artifact?: Artifact;
+  project: Project;
+  messageId: string;
+  code: string;
+  onSaveSuccess?: (artifact: Artifact) => void;
+}
+
+export function EditAppModal({
+  artifact,
+  project,
+  messageId,
+  code,
+  onSaveSuccess,
+  ...props
+}: Props) {
+  const { onRequestClose } = props;
+  const id = useId();
+  const {
+    setConfirmOnRequestClose,
+    clearConfirmOnRequestClose,
+    onRequestCloseSafe,
+  } = useModalControl();
+
+  const {
+    mutate: mutateSave,
+    isError: isSaveError,
+    error: saveError,
+  } = useSaveApp({ project, onSaveSuccess: () => onRequestClose() });
+
+  const editMode = artifact != undefined;
+
+  const {
+    control,
+    register,
+    handleSubmit,
+    formState: { errors, isValid, isSubmitting, dirtyFields },
+  } = useForm<FormValues>({
+    defaultValues: {
+      name: artifact?.name || '',
+    },
+    mode: 'onChange',
+  });
+
+  useEffect(() => {
+    if (!isEmpty(dirtyFields))
+      setConfirmOnRequestClose(
+        'Your app has unsaved changes, do you really want to leave?',
+      );
+    else clearConfirmOnRequestClose();
+    return () => clearConfirmOnRequestClose();
+  }, [clearConfirmOnRequestClose, dirtyFields, setConfirmOnRequestClose]);
```

Comment:
Almost identical code is inside UserToolModal, so we could maybe abstract it into separate hook. But it's not really necessary right now.